crates/bevy_ecs/src/schedule/executor_parallel.rs

```diff
@@ -124,40 +144,46 @@ impl ParallelSystemExecutor for ParallelExecutor {
             }
         }
 
-        ComputeTaskPool::init(TaskPool::default).scope(|scope| {
-            self.prepare_systems(scope, systems, world);
-            if self.should_run.count_ones(..) == 0 {
-                return;
-            }
-            let parallel_executor = async {
-                // All systems have been ran if there are no queued or running systems.
-                while 0 != self.queued.count_ones(..) + self.running.count_ones(..) {
-                    self.process_queued_systems();
-                    // Avoid deadlocking if no systems were actually started.
-                    if self.running.count_ones(..) != 0 {
-                        // Wait until at least one system has finished.
-                        let index = self
-                            .finish_receiver
-                            .recv()
-                            .await
-                            .unwrap_or_else(|error| unreachable!("{}", error));
-                        self.process_finished_system(index);
-                        // Gather other systems than may have finished.
-                        while let Ok(index) = self.finish_receiver.try_recv() {
+        let thread_executor = world.get_resource::<MainThreadExecutor>().map(|e| &*e.0);
+
+        ComputeTaskPool::init(TaskPool::default).scope_with_executor(
+            false,
+            thread_executor,
+            |scope| {
+                self.prepare_systems(scope, systems, world);
+                if self.should_run.count_ones(..) == 0 {
+                    return;
+                }
+                let parallel_executor = async {
+                    // All systems have been ran if there are no queued or running systems.
+                    while 0 != self.queued.count_ones(..) + self.running.count_ones(..) {
```

Comment:
```suggestion
                    while  self.queued.count_ones(..) + self.running.count_ones(..) > 0 {
```

I find this framing of the logic much clearer.

Reply:
Better yet, you can use `FixedBitset::is_clear` (NOT `is_empty`). It should be much faster due to not needing to use masks underneath.

I didn't comment on this due it mostly being an indentation change, but if it's going to be changed, might as well be a bit more readable (and performant).

Reply:
I don't think I should be changing this logic. The only change to this is indentation in this PR.

Reply:
I'm fine with that. Feel free to resolve this.